src/zacas.adoc

```diff
@@ -0,0 +1,260 @@
+aligned for _quadwords_, eight-byte aligned for _doublewords_, and four-byte
+aligned for _words_). And the same exception options apply if the address
+is not naturally aligned.
+
+Just as for AMOs in the A extension, the `AMOCAS.W/D/Q` optionally provide
+release consistency semantics, using the `aq` and `rl` bits, to help implement
+multiprocessor synchronization. The memory operation performed by an
+`AMOCAS.W/D/Q`, when successful, has acquire semantics if `aq` bit is 1 and has
+release semantics if `rl` bit is 1. The memory operation performed by an
+`AMOCAS.W/D/Q`, when not successful, has acquire semantics if `aq` bit is 1 but
+does not have release semantics, regardless of `rl`.
+
+A FENCE instruction may be used to order the memory read access and, if
+produced, the memory write access by an `AMOCAS.W/D/Q` instruction.
+
+[NOTE]
+====
+An unsuccessful `AMOCAS.W/D/Q` may either not perform a memory write or may
+write back the old value loaded from memory. The memory write, if produced, does
+not have release semantics, regardless of `rl`.
+====
+
+An `AMOCAS.W/D/Q` instruction always requires write permissions.
+
+<<<
+
+[NOTE]
+====
+The following example code sequence illustrates the use of `AMOCAS.Q` to
+implement the _enqueue_ operation for a non-blocking concurrent queue using the
+algorithm outlined in cite:[queue]. The algorithm atomically operates on a
+pointer and its associated modification counter using the `AMOCAS.Q` instruction
+to avoid the ABA problem.
+
+[listing]
+# Enqueue operation of a non-blocking concurrent queue.
+# Data structures used by the queue:
+#   structure pointer_t {ptr:   node_t *, count: uint64_t}
+#   structure node_t    {next: pointer_t, value: data type}
+#   structure queue_t   {Head: pointer_t, Tail:  pointer_t}
+# Inputs to the procedure:
+#   a0 - address of Tail variable
+#   a4 - address of a new node to insert at tail
+enqueue:
+  ld   a6, (a0)          # a6 = Tail.ptr
+  ld   a7, 8(a0)         # a7 = Tail.count
+  ld   a2, (a6)          # a2 = Tail.ptr->next.ptr
+  ld   a3, 8(a6)         # a3 = Tail.ptr->next.count
+  ld   t1, (a0)
+  ld   t2, 8(a0)
+  bne  a6, t1, enqueue   # Retry if Tail & next are not consistent
+  bne  a7, t2, enqueue   # Retry if Tail & next are not consistent
+  bne  a2, x0, move_tail # Was tail pointing to the last node?
+  mv   t1, a2            # Save Tail.ptr->next.ptr
+  mv   t2, a3            # Save Tail.ptr->next.count
+  addi a5, a3, 1         # Link the node at the end of the list
+  amocas.q.aqrl a2, a4, (a6)
+  bne  a2, t1, enqueue   # Retry if CAS failed
+  bne  a3, t2, enqueue   # Retry if CAS failed
+  addi a5, a7, 1         # Update Tail to the inserted node
+  amocas.q.aqrl a6, a4, (a0)
+  ret                    # Enqueue done
+move_tail:               # Tail was not pointing to the last node
+  addi a3, a7, 1         # Try to swing Tail to the next node
+  amocas.q.aqrl a6, a2, (a0)
+  j    enqueue           # Retry
+
+====
+
+=== Additional AMO PMAs
```

Comment:
This section should ideally be part of AMO PMA subsection in the Priv specifcation. 
When moved into the AMO PMAs section, we can remove the first sentence.  
@gfavor - what do you think?